### PR TITLE
Added trap command which will be running debug commands on exit

### PR DESCRIPTION
**Pull request by amar266**

### CODE REVIEW

**Comment by bodepd** [comment]
please test this

File: build_scripts/deploy.sh [hunk 1]
```diff
@@ -19,12 +19,14 @@ then
     nova list | grep test${BUILD_NUMBER} | cut -f2 -d' ' | while read uuid; do nova console-log $uuid | grep Giving.up.on.md && nova reboot $uuid || true; done
 fi
 
+trap "ssh -o ServerAliveInterval=30 -o UserKnownHostsFile=/dev/null -o StrictHostKeyChecking=no ${ssh_user:-jenkins}@${ip} python -m jiocloud.orchestrate debug_verify_hosts ${BUILD_NUMBER}" exit
```
Comment (bodepd) [inline build_scripts/deploy.sh]: this code uses the variable ip which has not yet been defined. It need to be moved to at least be after line 26.

**Comment by amar266** [comment]
Corrected the mistake and changed the function name

File: build_scripts/deploy.sh [hunk 2]
```diff
@@ -23,8 +23,10 @@ time $timeout 1200 bash -c 'while ! bash -c "ip=$(python -m jiocloud.utils get_i
...
 
+trap "ssh -o ServerAliveInterval=30 -o UserKnownHostsFile=/dev/null -o StrictHostKeyChecking=no ${ssh_user:-jenkins}@${ip} python -m jiocloud.orchestrate debug_timeout ${BUILD_NUMBER}" exit
+
 time $timeout 600 bash -c "while ! ssh -o ServerAliveInterval=30 -o UserKnownHostsFile=/dev/null -o StrictHostKeyChecking=no ${ssh_user:-jenkins}@${ip} python -m jiocloud.orchestrate trigger_update ${BUILD_NUMBER}; do sleep 5; done"
 
-time $timeout 3000 bash -c "while ! python -m jiocloud.apply_resources list --project_tag=${project_tag} environment/${layout}.yaml | sed -e 's/_/-/g' | ssh -o ServerAliveInterval=30 -o UserKnownHostsFile=/dev/null -o StrictHostKeyChecking=no ${ssh_user:-jenkins}@${ip} python -m jiocloud.orchestrate verify_hosts ${BUILD_NUMBER} ; do sleep 5; done"
+time $timeout 3000 bash -c "while ! python -m jiocloud.apply_resources list --project_tag=${project_tag} environment/${layout}.yaml | sed -e 's/_/-/g' | ssh -o ServerAliveInterval=30 -o UserKnownHostsFile=/dev/null -o StrictHostKeyChecking=no ${ssh_user:-jenkins}@${ip} python -m jiocloud.orchestrate verify_hosts ${BUILD_NUMBER} ; do sleep 5; done";
```
Comment (sorenh) [inline build_scripts/deploy.sh]: I don't think this change (the trailing semicolon) is necessary?

**Comment by sorenh** [comment]
please test this

**Comment by sorenh** [comment]
Oh, right, this depends on https://github.com/JioCloud/python-jiocloud/pull/42

**Comment by amar266** [comment]
Removed the semicolon and it depends on  JioCloud/python-jiocloud#42

File: build_scripts/deploy.sh [hunk 3]
```diff
@@ -23,6 +23,8 @@ time $timeout 1200 bash -c 'while ! bash -c "ip=$(python -m jiocloud.utils get_i
 
 ip=$(python -m jiocloud.utils get_ip_of_node ${consul_bootstrap_node:-bootstrap1}_${project_tag})
 
+trap "ssh -o ServerAliveInterval=30 -o UserKnownHostsFile=/dev/null -o StrictHostKeyChecking=no ${ssh_user:-jenkins}@${ip} python -m jiocloud.orchestrate debug_timeout ${BUILD_NUMBER}" exit
```
Comment (bodepd) [inline build_scripts/deploy.sh]: does this need to exit 1 to ensure it still shows up as a failure in jenkins?

Comment (sorenh) [inline build_scripts/deploy.sh]: Nope. Unless you explicitly exit N inside the trap command, the final return value of the script is intact.

**Comment by sorenh** [comment]
please test this

File: build_scripts/deploy.sh (hunk 3 above)
Comment (sorenh) [inline build_scripts/deploy.sh]: Hm. The test run suggests I'm wrong. Weird. I tested it locally. I must have gotten it wrong.

Comment (amar266) [inline build_scripts/deploy.sh]: I think exit sigspec will not work , it is running trap commands in all cases,instead we can use ERR sigspec.

Comment (sorenh) [inline build_scripts/deploy.sh]: No, it's fine to always run it.

**Comment by amar266** [comment]
please test this

**Comment by amar266** [comment]
please test this

**Comment by amar266** [comment]
please test this

**Pull request merged**